### PR TITLE
added *larf and *larfg LAPACK wrappers

### PR DESCRIPTION
**Pull request by gasagna**
I have added a couple of wrappers to LAPACK routines that were not currently implemented. Tests are included too.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Do we intend to use these in the LinearAlgebra API or is it mainly for completeness?

**Comment by StefanKarpinski** [comment]
This was my suggestion as a first step:

https://discourse.julialang.org/t/adding-more-lapack-wrappers-to-base-julia/25490

I'm not sure what these do, so I'm not sure if there's a sensible way to expose them at a higher level, but I figure as long as there are some tests to make sure they work and prevent them from bitrotting, there's no real harm in having them.

Also, "zlarf" is the best function name ever.

**Comment by ViralBShah** [comment]
The longer term goal is to make LAPACK less special to Julia (just like BLAS) allowing for native Julia implementations, stuff like FLAME, etc. But that is a post 2.0 discussion at this point, and not relevant to this thread. I am personally ok with adding more LAPACK wrappers in the meanwhile. The PR looks good too. While I have not verified the LAPACK arguments, passing tests is good enough.

It would be good for @andreasnoack to chime in and review.

**Review by vtjnash — COMMENTED** [review]
It doesn't matter...until it does https://developer.r-project.org/Blog/public/2019/05/15/gfortran-issues-with-lapack/index.html

File: stdlib/LinearAlgebra/src/lapack.jl [hunk 1]
```diff
@@ -753,6 +753,71 @@ function geqp3!(A::AbstractMatrix{<:BlasFloat})
...
+            m, n = size(C)
+            chkside(side)
+            ldc = max(1, stride(C, 2))
+            l = side == 'L' ? n : m
+            incv  = BlasInt(1)
+            ccall((@blasfunc($larf), liblapack), Cvoid,
+                (Ref{UInt8}, Ref{BlasInt}, Ref{BlasInt}, Ptr{$elty}, Ref{BlasInt},
+                 Ref{$elty}, Ptr{$elty}, Ref{BlasInt}, Ptr{$elty}),
+                side, m, n, v, incv,
+                τ, C, ldc, work)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/lapack.jl]:
```suggestion
                τ, C, ldc, work, 1)
```

**Comment by ViralBShah** [comment]
@vtjnash That should be a separate issue about fixing our build flags. Good to know about it, and it probably needs to go into BB.

**Comment by gasagna** [comment]
> Do we intend to use these in the LinearAlgebra API or is it mainly for completeness?

Mainly for completeness. There is https://github.com/JuliaMatrices/LAPACK.jl, but it's not very active. At this point, Julia's LAPACK wrappers are the most complete out there.

> I'm not sure what these do

They can be used to construct and apply Householder reflections. The LAPACK api is more complete (with blocked versions), but this is sufficient for my purposes...

**Comment by andreasnoack** [comment]
Just curious: the functionality of larf and larfg should be covered already by https://github.com/JuliaLang/julia/blob/21823899cc1724871a8e679287a94bb4acee72d7/stdlib/LinearAlgebra/src/generic.jl#L1274-L1295 and https://github.com/JuliaLang/julia/blob/21823899cc1724871a8e679287a94bb4acee72d7/stdlib/LinearAlgebra/src/generic.jl#L1298-L1322 so why did you need the LAPACK versions?

**Comment by gasagna** [comment]
I need to be able to apply householder reflectors from both right and left. Retrospectively I could have also implemented a pure julia function to apply reflectors from the right, based on the `reflectorApply!` function which applies reflectors from the left only.

**Comment by andreasnoack** [comment]
I'd forgotten the that the version that applies them from the right here. (I knew I'd been using one but it's in https://github.com/JuliaLinearAlgebra/GenericLinearAlgebra.jl/blob/a56a3ab1b568debd7bc77e0053582e54e4beed10/src/qr.jl#L17-L36)

**Comment by gasagna** [comment]
I guess it depends on if we want to have a complete eltype-generic API for linear algebra in Base at some point in the future that is competitive with LAPACK, or if that should go to a package (GenericLinearAlgebra.jl). In that case, I can start making a PR to move this function here, at least to have more functionality for Householder transformations.

**Pull request merged**